File: internal/saas/api/workspace_fingerprint.go
package api

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io/fs"
	"net/http"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strconv"
	"strings"
	"time"

	"github.com/sipeed/picoclaw/internal/saas/store"
	"github.com/sipeed/picoclaw/internal/saas/tenant"
	buildconfig "github.com/sipeed/picoclaw/pkg/config"
)

const (
	workspaceSyncStatusSynced   = "synced"
	workspaceSyncStatusDiverged = "diverged"
	workspaceSyncStatusUnknown  = "unknown"
)

var (
	workspaceFingerprintExcludeNames = map[string]struct{}{
		"sessions":                    {},
		"whatsapp":                    {},
		"state":                       {},
		"output":                      {},
		"logs":                        {},
		"cache":                       {},
		".cache":                      {},
		"node_modules":                {},
		".pnpm-store":                 {},
		"auth.json":                   {},
		"heartbeat.log":               {},
		"gerar_pdf_mamiferos.py":      {},
		"mamiferos_apresentacao.html": {},
		"RELATORIO-MELHORIAS.md":      {},
		".git":                        {},
		".vscode":                     {},
		".idea":                       {},
		"__pycache__":                 {},
	}
	workspaceFingerprintExcludeSuffixes = []string{".log", ".tmp.json", ".pid", ".sock", ".pyc"}
	workspaceSecurityTokenPattern       = regexp.MustCompile(`(?m)^(\s*-\s*)([A-Za-z][A-Za-z0-9_\-]{29,})\s*$`)
)

type workspaceContentFingerprint struct {
	HashSHA256 string
	FileCount  int
}

type workspaceBaselineManifest struct {
	HashSHA256 string
	FileCount  int
}

type workspaceSyncStatusResponse struct {
	WorkspaceID           string    `json:"workspace_id"`
	WorkspaceSlug         string    `json:"workspace_slug"`
	AdminHashSHA256       string    `json:"admin_hash_sha256"`
	AdminFileCount        int       `json:"admin_file_count"`
	DeployedGitHashSHA256 string    `json:"deployed_git_hash_sha256"`
	DeployedFileCount     int       `json:"deployed_file_count"`
	DeployedGitCommit     string    `json:"deployed_git_commit"`
	Status                string    `json:"status"`
	CheckedAt             time.Time `json:"checked_at"`
}

func (h *Handler) handleGetWorkspaceSyncStatus(w http.ResponseWriter, r *http.Request) {
	ws, ok := h.getWorkspace(w, r)
	if !ok {
		return
	}

	resp := workspaceSyncStatusForWorkspace(ws, time.Now().UTC())
	writeJSON(w, http.StatusOK, resp)
}

func workspaceSyncStatusForWorkspace(ws *store.Workspace, checkedAt time.Time) workspaceSyncStatusResponse {
	resp := workspaceSyncStatusResponse{
		WorkspaceID:       ws.ID,
		WorkspaceSlug:     ws.Slug,
		DeployedGitCommit: buildconfig.GitCommit,
		Status:            workspaceSyncStatusUnknown,
		CheckedAt:         checkedAt,
	}

	if _, ok := readEmbeddedBaselineManifest(); ok {
		if fp, err := fingerprintEmbeddedBaseline(); err == nil {
			resp.DeployedGitHashSHA256 = fp.HashSHA256
			resp.DeployedFileCount = fp.FileCount
		}
	}

	homeDir := filepath.Join(ws.HostPath, tenant.WorkspaceHomeSubdir)
	if fp, err := fingerprintWorkspaceHomeDir(homeDir); err == nil {
		resp.AdminHashSHA256 = fp.HashSHA256
		resp.AdminFileCount = fp.FileCount
	}

	if resp.AdminHashSHA256 == "" || resp.DeployedGitHashSHA256 == "" {
		return resp
	}
	if resp.AdminHashSHA256 == resp.DeployedGitHashSHA256 {
		resp.Status = workspaceSyncStatusSynced
	} else {
		resp.Status = workspaceSyncStatusDiverged
	}
	return resp
}

func readEmbeddedBaselineManifest() (workspaceBaselineManifest, bool) {
	data, err := baselineWorkspaceFS.ReadFile("baseline-workspace/SYNCED_FROM")
	if err != nil {
		return workspaceBaselineManifest{}, false
	}
	var out workspaceBaselineManifest
	for _, line := range strings.Split(string(data), "\n") {
		line = strings.TrimSpace(line)
		if strings.HasPrefix(line, "content_hash_sha256:") {
			out.HashSHA256 = strings.TrimSpace(strings.TrimPrefix(line, "content_hash_sha256:"))
			continue
		}
		if strings.HasPrefix(line, "file_count:") {
			n, err := strconv.Atoi(strings.TrimSpace(strings.TrimPrefix(line, "file_count:")))
			if err == nil {
				out.FileCount = n
			}
		}
	}
	if out.HashSHA256 == "" {
		return workspaceBaselineManifest{}, false
	}
	return out, true
}

func fingerprintEmbeddedBaseline() (workspaceContentFingerprint, error) {
	const embedRoot = "baseline-workspace"
	entries := map[string][]byte{}
	if err := fs.WalkDir(baselineWorkspaceFS, embedRoot, func(path string, d fs.DirEntry, walkErr error) error {
		if walkErr != nil {
			return walkErr
		}
		if path == embedRoot || d.IsDir() {
			return nil
		}
		rel := strings.TrimPrefix(path, embedRoot+"/")
		if rel == "SYNCED_FROM" {
			return nil
		}
		data, err := baselineWorkspaceFS.ReadFile(path)
		if err != nil {
			return fmt.Errorf("read embedded baseline %s: %w", rel, err)
		}
		entries[rel] = data
		return nil
	}); err != nil {
		return workspaceContentFingerprint{}, err
	}
	return hashWorkspaceFingerprintEntries(entries), nil
}

func fingerprintWorkspaceHomeDir(homeDir string) (workspaceContentFingerprint, error) {
	entries := map[string][]byte{}
	if err := filepath.WalkDir(homeDir, func(path string, d fs.DirEntry, walkErr error) error {
		if walkErr != nil {
			return walkErr
		}
		if path == homeDir {
			return nil
		}
		name := d.Name()
		if shouldSkipWorkspaceFingerprintName(name) {
			if d.IsDir() {
				return filepath.SkipDir
			}
			return nil
		}
		if d.Type()&fs.ModeSymlink != 0 {
			return nil
		}
		if d.IsDir() {
			return nil
		}
		rel, err := filepath.Rel(homeDir, path)
		if err != nil {
			return err
		}
		baselineRel, ok := homeRelToBaselineRel(rel)
		if !ok || shouldSkipWorkspaceFingerprintRel(baselineRel) {
			return nil
		}
		data, err := os.ReadFile(path)
		if err != nil {
			return err
		}
		entries[baselineRel] = data
		return nil
	}); err != nil {
		return workspaceContentFingerprint{}, err
	}
	if err := overlayEmbeddedBaselineSkippedFiles(entries); err != nil {
		return workspaceContentFingerprint{}, err
	}
	return hashWorkspaceFingerprintEntries(entries), nil
}

func homeRelToBaselineRel(rel string) (string, bool) {
	rel = filepath.ToSlash(filepath.Clean(rel))
	if rel == "." || rel == "" || rel == "SYNCED_FROM" {
		return "", false
	}
	if strings.HasPrefix(rel, tenant.WorkspaceHomeSubdir+"/") {
		rel = strings.TrimPrefix(rel, tenant.WorkspaceHomeSubdir+"/")
	}
	if strings.HasPrefix(rel, "workspace/") {
		return strings.TrimPrefix(rel, "workspace/"), true
	}
	if strings.Contains(rel, "/") {
		return "", false
	}
	if _, ok := embeddedBaselineHomeRootFiles[rel]; ok {
		return rel, true
	}
	return "", false
}

func overlayEmbeddedBaselineSkippedFiles(entries map[string][]byte) error {
	const embedRoot = "baseline-workspace"
	return fs.WalkDir(baselineWorkspaceFS, embedRoot, func(path string, d fs.DirEntry, walkErr error) error {
		if walkErr != nil {
			return walkErr
		}
		if path == embedRoot || d.IsDir() {
			return nil
		}
		rel := strings.TrimPrefix(path, embedRoot+"/")
		if rel == "SYNCED_FROM" {
			return nil
		}
		base := filepath.Base(rel)
		if base != "README.md" && base != ".gitkeep" {
			return nil
		}
		data, err := baselineWorkspaceFS.ReadFile(path)
		if err != nil {
			return fmt.Errorf("read embedded baseline overlay %s: %w", rel, err)
		}
		if _, exists := entries[rel]; !exists {
			entries[rel] = data
		}
		return nil
	})
}

func fingerprintWorkspaceSourceDir(workspaceDir string) (workspaceContentFingerprint, error) {
	entries := map[string][]byte{}
	if err := filepath.WalkDir(workspaceDir, func(path string, d fs.DirEntry, walkErr error) error {
		if walkErr != nil {
			return walkErr
		}
		if path == workspaceDir {
			return nil
		}
		name := d.Name()
		if shouldSkipWorkspaceFingerprintName(name) {
			if d.IsDir() {
				return filepath.SkipDir
			}
			return nil
		}
		if d.Type()&fs.ModeSymlink != 0 {
			return nil
		}
		if d.IsDir() {
			return nil
		}
		rel, err := filepath.Rel(workspaceDir, path)
		if err != nil {
			return err
		}
		rel = filepath.ToSlash(filepath.Clean(rel))
		if rel == "." || rel == "SYNCED_FROM" || shouldSkipWorkspaceFingerprintRel(rel) {
			return nil
		}
		data, err := os.ReadFile(path)
		if err != nil {
			return err
		}
		entries[rel] = normalizeWorkspaceSourceEntry(rel, data)
		return nil
	}); err != nil {
		return workspaceContentFingerprint{}, err
	}
	if err := overlayEmbeddedBaselineRootReadme(entries); err != nil {
		return workspaceContentFingerprint{}, err
	}
	return hashWorkspaceFingerprintEntries(entries), nil
}

func overlayEmbeddedBaselineRootReadme(entries map[string][]byte) error {
	data, err := baselineWorkspaceFS.ReadFile("baseline-workspace/README.md")
	if err != nil {
		return nil
	}
	entries["README.md"] = data
	return nil
}

func normalizeWorkspaceSourceEntry(rel string, data []byte) []byte {
	if strings.HasPrefix(rel, "memory/") && filepath.Base(rel) != ".gitkeep" {
		return workspaceMemoryStub(rel)
	}
	switch rel {
	case "config.json":
		return normalizeWorkspaceConfigJSONBytes(data)
	case ".security.yml":
		return workspaceSecurityTokenPattern.ReplaceAll(data, []byte("${1}REDACTED  # operator must replace with real key post-deploy"))
	default:
		return data
	}
}

func workspaceMemoryStub(rel string) []byte {
	ext := strings.ToLower(filepath.Ext(rel))
	switch ext {
	case ".md":
		stem := strings.TrimSuffix(filepath.Base(rel), filepath.Ext(rel))
		return []byte("# " + titleFromSlug(stem) + "\n\n")
	case ".json":
		return []byte("{}\n")
	default:
		return []byte{}
	}
}

func titleFromSlug(s string) string {
	parts := strings.Fields(strings.NewReplacer("-", " ", "_", " ").Replace(s))
	for i, part := range parts {
		if part == "" {
			continue
		}
		parts[i] = strings.ToUpper(part[:1]) + strings.ToLower(part[1:])
	}
	return strings.Join(parts, " ")
}

func normalizeWorkspaceConfigJSONBytes(data []byte) []byte {
	var raw map[string]any
	if err := json.Unmarshal(data, &raw); err != nil {
		return data
	}
	changed := false
	if agents, ok := raw["agents"].(map[string]any); ok {
		if defaults, ok := agents["defaults"].(map[string]any); ok {
			if _, has := defaults["workspace"]; has && defaults["workspace"] != "/root/.picoclaw/workspace" {
				defaults["workspace"] = "/root/.picoclaw/workspace"
				changed = true
			}
		}
		if list, ok := agents["list"].([]any); ok {
			for _, item := range list {
				agent, ok := item.(map[string]any)
				if !ok {
					continue
				}
				ws, ok := agent["workspace"].(string)
				if !ok {
					continue
				}
				if strings.HasPrefix(ws, "C:") || strings.HasPrefix(ws, "/Users") || strings.HasPrefix(ws, "/home/") {
					agent["workspace"] = "/root/.picoclaw/workspace"
					changed = true
				}
			}
		}
	}
	if list, ok := raw["model_list"].([]any); ok {
		for _, item := range list {
			model, ok := item.(map[string]any)
			if !ok {
				continue
			}
			if _, has := model["api_keys"]; has {
				model["api_keys"] = []any{"${LITELLM_KEY}"}
				changed = true
			}
			if _, has := model["api_key"]; has {
				delete(model, "api_key")
				model["api_keys"] = []any{"${LITELLM_KEY}"}
				changed = true
			}
		}
	}
	if !changed {
		return data
	}
	out, err := json.MarshalIndent(raw, "", "  ")
	if err != nil {
		return data
	}
	return append(out, '\n')
}

func hashWorkspaceFingerprintEntries(entries map[string][]byte) workspaceContentFingerprint {
	rels := make([]string, 0, len(entries))
	for rel := range entries {
		rel = filepath.ToSlash(filepath.Clean(rel))
		if rel == "." || rel == "SYNCED_FROM" {
			continue
		}
		rels = append(rels, rel)
	}
	sort.Strings(rels)

	h := sha256.New()
	count := 0
	for _, rel := range rels {
		data := entries[rel]
		h.Write([]byte(rel))
		h.Write([]byte{0})
		h.Write([]byte(strconv.Itoa(len(data))))
		h.Write([]byte{0})
		h.Write(data)
		h.Write([]byte{0})
		if rel != "README.md" {
			count++
		}
	}
	return workspaceContentFingerprint{
		HashSHA256: hex.EncodeToString(h.Sum(nil)),
		FileCount:  count,
	}
}

func shouldSkipWorkspaceFingerprintRel(rel string) bool {
	rel = filepath.ToSlash(filepath.Clean(rel))
	for _, part := range strings.Split(rel, "/") {
		if shouldSkipWorkspaceFingerprintName(part) {
			return true
		}
	}
	return false
}

func shouldSkipWorkspaceFingerprintName(name string) bool {
	if _, ok := workspaceFingerprintExcludeNames[name]; ok {
		return true
	}
	for _, suffix := range workspaceFingerprintExcludeSuffixes {
		if strings.HasSuffix(name, suffix) {
			return true
		}
	}
	return false
}

File: scripts/workspace-fingerprint.py
#!/usr/bin/env python3
"""Read-only workspace fingerprint helper.

Computes the same deterministic SHA256 shape used by
scripts/sync-baseline-workspace.py without writing to workspace/ or
internal/saas/api/baseline-workspace/.
"""
from __future__ import annotations

import argparse
import hashlib
import json
import re
import sys
from pathlib import Path


REPO_ROOT = Path(__file__).resolve().parents[1]
DEFAULT_WORKSPACE = REPO_ROOT / "workspace"
DEFAULT_BASELINE = REPO_ROOT / "internal" / "saas" / "api" / "baseline-workspace"
DEFAULT_MANIFEST = DEFAULT_BASELINE / "SYNCED_FROM"

EXCLUDE_NAMES_EXACT = {
    "sessions",
    "whatsapp",
    "state",
    "output",
    "logs",
    "cache",
    ".cache",
    "node_modules",
    ".pnpm-store",
    "auth.json",
    "heartbeat.log",
    "gerar_pdf_mamiferos.py",
    "mamiferos_apresentacao.html",
    "RELATORIO-MELHORIAS.md",
    ".git",
    ".vscode",
    ".idea",
    "__pycache__",
}
EXCLUDE_SUFFIXES = (".log", ".tmp.json", ".pid", ".sock", ".pyc")
HOME_ROOT_FILES = {
    ".security.yml",
    "auth.json",
    "config.json",
    "launcher_policy.json",
    "ui-visibility.json",
    "ui.config.fields.json",
}
SECURITY_TOKEN_RE = re.compile(
    r"^(\s*-\s*)([A-Za-z][A-Za-z0-9_\-]{29,})\s*$",
    re.MULTILINE,
)


def should_skip_name(name: str) -> bool:
    return name in EXCLUDE_NAMES_EXACT or name.endswith(EXCLUDE_SUFFIXES)


def should_skip_rel(rel: str) -> bool:
    return any(should_skip_name(part) for part in Path(rel).parts)


def title_from_slug(stem: str) -> str:
    return stem.replace("-", " ").replace("_", " ").title()


def memory_stub(rel: str) -> bytes:
    path = Path(rel)
    suffix = path.suffix.lower()
    if suffix == ".md":
        return f"# {title_from_slug(path.stem)}\n\n".encode("utf-8")
    if suffix == ".json":
        return b"{}\n"
    return b""


def normalize_config_json_bytes(data: bytes) -> bytes:
    try:
        raw = json.loads(data.decode("utf-8"))
    except Exception:
        return data

    changed = False
    defaults = raw.get("agents", {}).get("defaults", {})
    if "workspace" in defaults and defaults["workspace"] != "/root/.picoclaw/workspace":
        defaults["workspace"] = "/root/.picoclaw/workspace"
        changed = True

    for agent in raw.get("agents", {}).get("list", []) or []:
        ws = agent.get("workspace") if isinstance(agent, dict) else None
        if isinstance(ws, str) and (
            ws.startswith("C:") or ws.startswith("/Users") or ws.startswith("/home/")
        ):
            agent["workspace"] = "/root/.picoclaw/workspace"
            changed = True

    placeholder_plural = ["${LITELLM_KEY}"]
    for model in raw.get("model_list", []) or []:
        if not isinstance(model, dict):
            continue
        if "api_keys" in model and model["api_keys"] != placeholder_plural:
            model["api_keys"] = placeholder_plural
            changed = True
        if "api_key" in model:
            del model["api_key"]
            model["api_keys"] = placeholder_plural
            changed = True

    if not changed:
        return data
    return (json.dumps(raw, indent=2, ensure_ascii=False) + "\n").encode("utf-8")


def normalize_repo_entry(rel: str, data: bytes) -> bytes:
    if rel.startswith("memory/") and Path(rel).name != ".gitkeep":
        return memory_stub(rel)
    if rel == "config.json":
        return normalize_config_json_bytes(data)
    if rel == ".security.yml":
        return SECURITY_TOKEN_RE.sub(
            r"\1REDACTED  # operator must replace with real key post-deploy",
            data.decode("utf-8", errors="replace"),
        ).encode("utf-8")
    return data


def iter_files(root: Path):
    for path in sorted(root.rglob("*")):
        if path.is_dir():
            continue
        rel = path.relative_to(root).as_posix()
        if should_skip_rel(rel):
            continue
        yield rel, path


def build_repo_entries(source: Path, baseline_root: Path) -> dict[str, bytes]:
    if not source.is_dir():
        raise FileNotFoundError(f"workspace source not found: {source}")
    entries: dict[str, bytes] = {}
    for rel, path in iter_files(source):
        if rel == "SYNCED_FROM":
            continue
        entries[rel] = normalize_repo_entry(rel, path.read_bytes())

    # baseline-workspace/README.md is intentionally preserved at the embedded
    # baseline root. It is not copied from workspace/, but the manifest hash
    # includes it, so include the current baseline copy for comparison.
    root_readme = baseline_root / "README.md"
    if root_readme.is_file():
        entries["README.md"] = root_readme.read_bytes()
    return entries


def build_baseline_entries(source: Path, _baseline_root: Path) -> dict[str, bytes]:
    if not source.is_dir():
        raise FileNotFoundError(f"baseline source not found: {source}")
    entries: dict[str, bytes] = {}
    for rel, path in iter_files(source):
        if rel == "SYNCED_FROM":
            continue
        entries[rel] = path.read_bytes()
    return entries


def home_rel_to_baseline_rel(rel: str) -> str | None:
    rel = Path(rel).as_posix()
    if rel.startswith("home/"):
        rel = rel.removeprefix("home/")
    if rel.startswith("workspace/"):
        return rel.removeprefix("workspace/")
    if "/" in rel:
        return None
    if rel in HOME_ROOT_FILES:
        return rel
    return None


def build_home_entries(source: Path, baseline_root: Path) -> dict[str, bytes]:
    if not source.is_dir():
        raise FileNotFoundError(f"home source not found: {source}")
    entries: dict[str, bytes] = {}
    for rel, path in iter_files(source):
        mapped = home_rel_to_baseline_rel(rel)
        if not mapped or mapped == "SYNCED_FROM" or should_skip_rel(mapped):
            continue
        entries[mapped] = path.read_bytes()

    # README.md and .gitkeep files are embedded in the deployed baseline hash
    # but not materialized by extractEmbeddedBaseline(). Overlay them from the
    # same deployed baseline so a freshly seeded admin workspace compares OK.
    if baseline_root.is_dir():
        for rel, path in iter_files(baseline_root):
            if rel == "SYNCED_FROM":
                continue
            if Path(rel).name in {"README.md", ".gitkeep"} and rel not in entries:
                entries[rel] = path.read_bytes()
    return entries


def hash_entries(entries: dict[str, bytes]) -> tuple[str, int]:
    h = hashlib.sha256()
    count = 0
    for rel in sorted(k for k in entries if k != "SYNCED_FROM"):
        data = entries[rel]
        h.update(rel.encode("utf-8"))
        h.update(b"\0")
        h.update(str(len(data)).encode("ascii"))
        h.update(b"\0")
        h.update(data)
        h.update(b"\0")
        if rel != "README.md":
            count += 1
    return h.hexdigest(), count


def read_manifest(path: Path) -> tuple[str | None, int | None]:
    if not path.is_file():
        return None, None
    digest: str | None = None
    count: int | None = None
    for line in path.read_text(encoding="utf-8").splitlines():
        line = line.strip()
        if line.startswith("content_hash_sha256:"):
            digest = line.split(":", 1)[1].strip()
        elif line.startswith("file_count:"):
            try:
                count = int(line.split(":", 1)[1].strip())
            except ValueError:
                count = None
    return digest, count


def main() -> int:
    parser = argparse.ArgumentParser(description="Compute a read-only workspace fingerprint.")
    parser.add_argument("--source", type=Path, default=DEFAULT_WORKSPACE)
    parser.add_argument("--layout", choices=("repo", "baseline", "home"), default="repo")
    parser.add_argument("--baseline-root", type=Path, default=DEFAULT_BASELINE)
    parser.add_argument("--manifest", type=Path, default=DEFAULT_MANIFEST)
    parser.add_argument("--compare-manifest", action="store_true")
    args = parser.parse_args()

    builders = {
        "repo": build_repo_entries,
        "baseline": build_baseline_entries,
        "home": build_home_entries,
    }
    try:
        entries = builders[args.layout](args.source, args.baseline_root)
        digest, count = hash_entries(entries)
    except Exception as exc:
        print(f"status: unknown")
        print(f"error: {exc}", file=sys.stderr)
        return 2

    print(f"source: {args.source}")
    print(f"layout: {args.layout}")
    print(f"content_hash_sha256: {digest}")
    print(f"file_count: {count}")

    if not args.compare_manifest:
        return 0

    expected_digest, expected_count = read_manifest(args.manifest)
    print(f"manifest: {args.manifest}")
    print(f"manifest_hash_sha256: {expected_digest or ''}")
    print(f"manifest_file_count: {expected_count if expected_count is not None else ''}")

    if not expected_digest:
        print("status: unknown")
        return 2
    if digest == expected_digest:
        print("status: synced")
        return 0
    print("status: diverged")
    return 1


if __name__ == "__main__":
    raise SystemExit(main())
